### PR TITLE
Add labels to OperationMetrics and add Specs to observation package

### PR DESCRIPTION
**Pull request by efritz**
Add `NewSpecs` to the observation package, which reduces the boilerplate necessary in the implementation of observed structs.

See a sample usage [here](https://github.com/sourcegraph/sourcegraph/pull/10361/files#diff-2ef42b88d2d86ceddbdf982f5681e8ceR14).

### CODE REVIEW

File: internal/metrics/operation.go [hunk 1]
```diff
@@ -58,13 +59,19 @@ func WithErrorsHelp(text string) OperationMetricsOption {
...
-// NewOperationMetrics creates an OperationMetrics value. The supplied operationName should
+// WithLabels overrides the default labels for all metrics.
+func WithLabels(labels ...string) OperationMetricsOption {
+	return func(o *operationMetricOptions) { o.labels = labels }
+}
+
+// NewOperationMetrics creates an OperationMetrics value. The supplied metricName should
 // be underscore_cased as it is used in the metric name.
-func NewOperationMetrics(subsystem, metricPrefix, operationName string, fns ...OperationMetricsOption) *OperationMetrics {
+func NewOperationMetrics(subsystem, metricName string, fns ...OperationMetricsOption) *OperationMetrics {
```
Comment (slimsag) [inline internal/metrics/operation.go:69]: It's still a prefix, so I think you should use `metricPrefix` (also in the docstring) instead of `metricName`

Comment (efritz) [reply internal/metrics/operation.go]: Updated.

File: internal/observation/observation.go [hunk 2]
```diff
@@ -108,3 +110,73 @@ func With(ctx context.Context, args Args) (context.Context, FinishFn) {
...
+// an operation. For usage details, see the With function in this package.
+func (s Specs) With(
+	ctx context.Context,
+	err *error,
+	operationName string,
+	preFields ...log.Field,
+) (context.Context, FinishFn) {
+	spec, ok := s.specs[operationName]
+	if !ok {
+		return ctx, func(count float64, additionalLogFields ...log.Field) {}
```
Comment (slimsag) [inline internal/observation/observation.go:169]: This means in your code if you accidently typo the operation name, tracing, logging, and metrics will not work?

Comment (efritz) [reply internal/observation/observation.go:169]: There's not a good way to statically determine different operation shapes. We could instead of doing nothing we could insert an error in the flow or log a warning. Thoughts on those solutions?

File: internal/observation/observation.go [hunk 3]
```diff
@@ -108,3 +110,73 @@ func With(ctx context.Context, args Args) (context.Context, FinishFn) {
...
+type Specs struct {
+	logger  logging.ErrorLogger
+	tracer  trace.Tracer
+	specs   map[string]Spec
+	metrics []*metrics.OperationMetrics
+}
+
+// NewSpecs creates a new Specs object that can be used with an observed struct. Before use,
+// the metrics in each Spec must be registered in a Prometheus registry via MustRegister.
+func NewSpecs(logger logging.ErrorLogger, tracer trace.Tracer, specs map[string]Spec) Specs {
```
Comment (slimsag) [inline internal/observation/observation.go:132]: I'm not convinced this is a good API. It feels like a secondary layer of indirection/abstraction on top of `observation.With` that should not be necessary and is not entirely clear how to make use of, and introduces questions like:

- Do I use `observation.NewSpecs`, or `observation.With`?
- What is a "spec"?
- I looked at the API and it's very confusing to me, even with examples.

I have watched Tomas's gophercon talk and this feels to me exactly like a move in the wrong direction described in that talk. However, I am convinced `observation.With` is a good idea because:

1. We do want all code to have proper logging, tracing, and metrics. We don't want any of these to be optional in practice, so not using a decorator pattern and instead having an abstraction which aims to enforce implementing all three does make sense.
2. It can handle that deduplication and prescribe easier ways to setup logging, tracing, and metrics.

But, the fact that you need this Specs abstraction on top just tells me one thing: the original `observation.With` API is fundamentally broken/wrong in some way at performing this abstraction and it is what we should revisit instead of adding on top of it.

Comment (slimsag) [reply internal/observation/observation.go:132]: From what I see, you have the following problems using `observation.With` which you've solved with this `Specs` approach:

1. You need to pass around `logger`, `metrics`, and `tracer` everywhere.
2. You need to pass the trace name / log name / metric labels / log fields at the usage site.
3. You need to pre-register Prometheus metrics

I can see why 1 & 3 are PITAs and worth solving, but the latter seems not as big a problem.

I would encourage you to adopt a more declarative API as follows:

1. Something that carries context:

```
observe := observation.Context(logger, tracer) // returns *observation.Context
```

2. Something that declares an operation to be observed and registers metrics with Prometheus:

```
// registers metrics with Prometheus
dbOperation := observe.Operation(observation.Op{
    Subsystem: subsystem,
    Metric: metrics.WithLabels("op"), // TODO: I am not actually sure a metrics package should exist
})
```

3. Something that records actions of that operation:

```
dbOperation.Record(observation.Record{
    LogName: "DB.GetUploadByID",
    TraceName: "db.get-upload-by-id",
    MetricLabels: []string{"GetUploadByID"},
})
```

Basically, I am begging / pleading for you to consider the end-to-end experience here. It seems clear to me we want an abstraction that enforces a standard usage of logging, metrics, and tracing all in one -- but that API should be as dead simple as it can be and likely should only involve using one package: the `observation` package.
